creator-node/src/routes/nodeSync.js

```diff
@@ -271,121 +298,67 @@ async function _nodesync (req, walletPublicKeys, creatorNodeEndpoint, dbOnlySync
+        latestBlockNumber: fetchedLatestBlockNumber,
+        clock: fetchedLatestClockVal,
+        clockRecords: fetchedClockRecords
+      } = fetchedCNodeUser
+
+      // Error if returned data is not within requested range
+      if (fetchedLatestClockVal < localMaxClockVal) {
+        throw new Error(`Cannot sync for localMaxClockVal ${localMaxClockVal} - imported data has max clock val ${fetchedLatestClockVal}`)
+      } else if (fetchedLatestClockVal === localMaxClockVal) {
+        // Already up to date, no sync necessary
+        req.logger.info(redisKey, `User ${fetchedWalletPublicKey} already up to date! Both nodes have latest clock value ${localMaxClockVal}`)
+        continue
+      } else if (localMaxClockVal !== -1 && fetchedClockRecords[0] && fetchedClockRecords[0].clock !== localMaxClockVal + 1) {
+        throw new Error(`Cannot sync - imported data is not contiguous. Local max clock val = ${localMaxClockVal} and imported min clock val ${fetchedClockRecords[0].clock}`)
+      }
+
+      // All DB updates must happen in single atomic tx - partial state updates will lead to data loss
       const transaction = await models.sequelize.transaction()
 
       try {
-        const cnodeUser = await models.CNodeUser.findOne({
-          where: { walletPublicKey: fetchedWalletPublicKey },
-          transaction
-        })
-        const fetchedLatestBlockNumber = fetchedCNodeUser.latestBlockNumber
-        const fetchedLatestClockVal = fetchedCNodeUser.clock
-
-        // Delete any previously stored data for cnodeUser in reverse table dependency order (cannot be parallelized).
-        if (cnodeUser) {
-          // Ensure imported data has higher blocknumber than already stored.
-          const latestBlockNumber = cnodeUser.latestBlockNumber
-          const latestClockValue = cnodeUser.clock
-
-          if (latestClockValue > fetchedLatestClockVal) {
-            throw new Error(`Imported data is outdated, will not sync. Imported latestBlockNumber \
-              ${fetchedLatestBlockNumber} Self latestBlockNumber ${latestBlockNumber}. \
-              fetched latestClockVal: ${fetchedLatestClockVal}, self latestClockVal: ${latestClockValue}`)
-          } else if (latestClockValue === fetchedLatestClockVal) {
-            // Already to update, no sync necessary
-            req.logger.info(redisKey, `User ${fetchedWalletPublicKey} already up to date! fetchedLatestClockVal=${fetchedLatestClockVal}, latestClockValue=${latestClockValue}`)
-            // the transaction declared outside the try/catch needs to be closed. if we call the continue
-            // and do not end the tx, it will never be closed
-            transaction.rollback()
-            continue
-          }
-
-          const cnodeUserUUID = cnodeUser.cnodeUserUUID
-          req.logger.info(redisKey, `beginning delete ops for cnodeUserUUID ${cnodeUserUUID}`)
-
-          const numAudiusUsersDeleted = await models.AudiusUser.destroy({
```

Comment:
i feel like this code is super useful. maybe we should break it out into a new function that takes in a tx and just like calls delete for a user state but no need to have it in the main nodesync flow anymore

Reply:
good idea - i added this back as an unused dbManager function, although it does not have a test case and i'm not gonna write one until everything else is done